gnc-fq-dump Err msg if AlphaVantage API Key is needed but missing

goodvibes2 opened the pull request:
As mentioned in Bug 796993.
Note Finance::Quote will also give an err msg but it will only show when gnc-fq-dump is run with -v for a stock quote.

Christopher Lam has also contributed scheme code, which I have included here, to give an error message during Get Quotes processing in Price Database if a currency quote or a stock quote for source 'alphavantage' or 'vanguard' is detected and the API key is missing. No warning message is given from 'Get Quotes' for the 'multiple source' sources that include alphavantage (canada,nyse,nasdaq + usa) as one of the other sources may work. Unlike previously, if a missing api key is detected, no further sources will be processed. I don't think this will be a problem as it is easy to get an api key.

fellen commented on libgnucash/quotes/gnc-fq-dump.in:
```diff
@@ -207,6 +218,7 @@ gnc-fq-dump	- Print out data from the F::Q module
     gnc-fq-dump europe 48406.PA 13000.PA
     gnc-fq-dump vwd 632034
     gnc-fq-dump ftportfolios FKYGTX
+    gnc-fq-dump currency USD AUD
```
I would put it at the beginning of the list as I estimate it globally seen as the most often used source.

goodvibes2 replied:
Thanks @fellen. I have moved the currency example to the front and added a couple more.

christopherlam commented:
Or here's a simpler patch. Like yours, it just checks for presence of the key; does not check whether it is accepted.

```patch
modified   gnucash/price-quotes.scm
@@ -130,6 +130,9 @@
                (gnc:debug "handling-request: " request)
                ;; we need to display the first element (the method,
                ;; so it won't be quoted) and then write the rest
+               (if (and (memq (car request) '(currency alphavantage))
+                        (string-null? (getenv "ALPHAVANTAGE_API_KEY")))
+                   (throw 'need-alphavantage-key))
                (with-output-to-port (fdes->outport (gnc-process-get-fd quoter 0))
                  (lambda ()
                    (display #\()
@@ -451,6 +454,11 @@ Run 'gnc-fq-update' as root to install them.")))
       (set! keep-going? #f)
       (show-error (N_ "There was a system error while retrieving the price quotes.")))
 
+     ((eq? (car fq-results) 'need-alphavantage-key)
+      (set! keep-going? #f)
+      (show-error (N_ "ERROR: ALPHAVANTAGE_API_KEY must be set for currency \
+quotes; get an API key at https://www.alphavantage.co")))
+
      ((not (list? (car fq-results)))
       (set! keep-going? #f)
       (show-error (N_ "There was an unknown error while retrieving the price quotes.")))
```

goodvibes2 commented:
Thanks @christopherlam :-)
Shall I put that in to my PR (as a separate commit)? I want to redo my commits anyway as I want to change the link to be to the wiki FAQ (as that says where to add it in Preferences) and also move example in synopsis and add some more examples.

christopherlam commented:
Go ahead 😉

goodvibes2 commented:
Hi @christopherlam ,
It doesn't give the error msg for a stock with source 'alphavantage', just says 1 failed...
Any idea what is the easiest way to get a currency quote requested? All my accounts/stocks are in AUD. I'll try setting up IBM in USD.

christopherlam commented:
I don't use fq regularly, not sure. This is a better fix for the error handler, and should be earlier in the error handling code.

```patch
modified   gnucash/price-quotes.scm
@@ -130,6 +130,9 @@
                (gnc:debug "handling-request: " request)
                ;; we need to display the first element (the method,
                ;; so it won't be quoted) and then write the rest
+               (if (and (memq (car request) '(currency alphavantage))
+                        (string-null? (getenv "ALPHAVANTAGE_API_KEY")))
+                   (throw 'need-alphavantage-key))
                (with-output-to-port (fdes->outport (gnc-process-get-fd quoter 0))
                  (lambda ()
                    (display #\()
@@ -451,6 +454,11 @@ Run 'gnc-fq-update' as root to install them.")))
       (set! keep-going? #f)
       (show-error (N_ "There was a system error while retrieving the price quotes.")))
 
+     ((memq 'need-alphavantage-key fq-results)
+      (set! keep-going? #f)
+      (show-error (N_ "ERROR: ALPHAVANTAGE_API_KEY must be set for currency \
+quotes; get an API key at https://www.alphavantage.co")))
+
      ((not (list? (car fq-results)))
       (set! keep-going? #f)
       (show-error (N_ "There was an unknown error while retrieving the price quotes.")))
```

goodvibes2 commented:
I set up stock IBM (qty 10) under 'Brokerage USD' using source NYSE and Get Quotes gets a price OK but no warning about KEY not being set for the other stock with source 'alphavantage' and no currency shows up in Price Database...
I took your mods out, and Get Quotes does get a USD Currency now.

goodvibes2 commented:
Still no good I'm afraid, don't get the error msg re AV for either currency or stock with alphavantage source. Just both of those failed:
Unable to retrieve quotes for these items:
ASX:AMP
CURRENCY:USD
Continue using only the good quotes?

Here's my diff in case I'm doing something stupid:
--- a/gnucash/price-quotes.scm
+++ b/gnucash/price-quotes.scm
@@ -130,6 +130,9 @@
                (gnc:debug "handling-request: " request)
                ;; we need to display the first element (the method,
                ;; so it won't be quoted) and then write the rest
+               (if (and (memq (car request) '(currency alphavantage))
+                        (string-null? (getenv "ALPHAVANTAGE_API_KEY")))
+                   (throw 'need-alphavantage-key))
                (with-output-to-port (fdes->outport (gnc-process-get-fd quoter 0))
                  (lambda ()
                    (display #\()
@@ -451,6 +454,11 @@ Run 'gnc-fq-update' as root to install them.")))
       (set! keep-going? #f)
       (show-error (N_ "There was a system error while retrieving the price quotes.")))
 
+     ((memq 'need-alphavantage-key fq-results)
+      (set! keep-going? #f)
+      (show-error (N_ "ERROR: ALPHAVANTAGE_API_KEY must be set for currency \
+quotes; see https://wiki.gnucash.org/wiki/FAQ#Q:_Why_doesn.27t_online_quoting_work.3F")))
+
      ((not (list? (car fq-results)))
       (set! keep-going? #f)
       (show-error (N_ "There was an unknown error while retrieving the price quotes.")))

That's it for me tonight - the dogs are hungry. :-)

fellen commented:
> I want to change the link to be to the wiki FAQ (as that says where to add it in Preferences)

If we have the theme already in the docs, that is preferred over the wiki. Potential targets:
https://code.gnucash.org/docs/C/gnucash-help/set-prefs.html#prefs-quotes
https://code.gnucash.org/docs/C/gnucash-help/fq-sources.html
https://code.gnucash.org/docs/C/gnucash-guide/invest-stockprice1.html#invest-stockprice-auto2
https://code.gnucash.org/docs/C/gnucash-guide/currency_manual.html#currency_howto1

Verify that the section ID is defined in the program: https://wiki.gnucash.org/wiki/Documentation_Update_Instructions#Telling_the_Program_of_a_New_Help_Context

fellen commented:
One problem: alphavantage can be used for the following sources:
```
sub methods {
    return ( alphavantage => \&alphavantage,
             canada       => \&alphavantage,
             usa          => \&alphavantage,
             nyse         => \&alphavantage,
             nasdaq       => \&alphavantage,
             vanguard     => \&alphavantage,
    );
}

```
Some of them are  'Multiple sources' like canada, usa, but we should add the other three to list for a warning.

goodvibes2 commented:
> One problem: alphavantage can be used for the following sources:
> 
> ```
> sub methods {
>     return ( alphavantage => \&alphavantage,
>              canada       => \&alphavantage,
>              usa          => \&alphavantage,
>              nyse         => \&alphavantage,
>              nasdaq       => \&alphavantage,
>              vanguard     => \&alphavantage,
>     );
> }
> ```
> 
> Some of them are 'Multiple sources' like canada, usa, but we should add the other three to the list for a warning.

Thanks for bringing this up Frank. I believe canada, usa, nyse and nasdaq are all multi sources.
I'll change gnc-fq-dump to give a warning for these, and also die for vanguard.

@christopherlam , can you please also give error message for source 'vanguard'?

christopherlam commented:
> can you please also give error message for source 'vanguard'?

all you'd do is add vanguard to the list `(memq (car request) '(currency alphavantage vanguard))` in the first part of the patch.

goodvibes2 commented:
Hi @christopherlam ,
I don't want to hassle you because I know there is a lot happening at the moment, but did you see my comments that mods to price-quotes.scm are not working?
I can create a test database for you if you wish?

christopherlam commented:
Yes please for datafile -- I don't use quotes myself.

From my pov both currency quotes (eg EUR GBP AUD) and alphavantage (TSLA AAPL) are working well.

goodvibes2 commented:
[Bug796993FqWarn.gnucash.zip](https://github.com/Gnucash/gnucash/files/4714352/Bug796993FqWarn.gnucash.zip)

What do you mean it is working well? The error msg is showing when api key is not set?

goodvibes2 commented:
@christopherlam  Could it be not working for me because I also have GC 3.4 installed in another location under $HOME?

goodvibes2 commented:
> > I want to change the link to be to the wiki FAQ (as that says where to add it in Preferences)
> 
> If we have the theme already in the docs, that is preferred over the wiki. Potential targets:
> https://code.gnucash.org/docs/C/gnucash-help/set-prefs.html#prefs-quotes
> https://code.gnucash.org/docs/C/gnucash-help/fq-sources.html
> https://code.gnucash.org/docs/C/gnucash-guide/invest-stockprice1.html#invest-stockprice-auto2
> https://code.gnucash.org/docs/C/gnucash-guide/currency_manual.html#currency_howto1
> 
> Verify that the document's section ID is defined in the program: https://wiki.gnucash.org/wiki/Documentation_Update_Instructions#Telling_the_Program_of_a_New_Help_Context

Hi @fellen ,
IIRC, those links are for the live documentation, and it could be up to 3 months before they get updated. I'd prefer to link to https://wiki.gnucash.org/wiki/FAQ#Q:_Why_doesn.27t_online_quoting_work.3F
as FQ can be volatile and the wiki can be quickly updated. What do you think?

christopherlam commented:
Here's a better patch. It works when ALPHAVANTAGE is empty, but will not replicate the previous behaviour to retrieve partial quotes. It's too difficult to do so.

```patch
modified   gnucash/price-quotes.scm
@@ -128,6 +128,9 @@
            (catch #t
              (lambda ()
                (gnc:debug "handling-request: " request)
+               (and (member (car request) '("currency" "alphavantage" "vanguard"))
+                    (not (getenv "ALPHAVANTAGE_API_KEY"))
+                    (throw 'need-alphavantage-key))
                ;; we need to display the first element (the method,
                ;; so it won't be quoted) and then write the rest
                (with-output-to-port (fdes->outport (gnc-process-get-fd quoter 0))
@@ -447,6 +450,11 @@
       (show-error (N_ "You are missing some needed Perl libraries.
 Run 'gnc-fq-update' as root to install them.")))
 
+     ((memq 'need-alphavantage-key fq-results)
+      (set! keep-going? #f)
+      (show-error (N_ "ERROR: ALPHAVANTAGE_API_KEY must be set for currency \
++quotes; get an API key at https://www.alphavantage.co")))
+
      ((memq 'system-error fq-results)
       (set! keep-going? #f)
       (show-error (N_ "There was a system error while retrieving the price quotes.")))
```

christopherlam commented:
> @christopherlam Could it be not working for me because I also have GC 3.4 installed in another location under $HOME?

No, my patch was simply incorrect. `(car request)` would be the string "currency" rather than the symbol 'currency, and getenv would return `#f` instead of `""`when the env is missing.

goodvibes2 commented:
Hi @christopherlam , price-quotes.scm is giving error msg now when api key is not set, but it is also giving the error when it shouldn't be needed. I unticked the 'Get Quotes' field on all but 1 security which had source ASX and it still gave the error. Unless I misunderstand and it DOES need to get a currency quote, this will force people to get an API key who previously have not had to, especially as IIUC  it now doesn't continue to process other requests.

christopherlam commented:
I don't know. I'm not familiar with how fq works. Try with/without `"currency"` in the patch. If you can summarise various sources -> fq sources let me know.

goodvibes2 commented:
Without "currency", it no longer gets the missing API key error, but (surprisingly to me) still tries to get a currency conversion rate but fails because API key is not set. Error:
Unable to retrieve quotes for these items:
 CURRENCY:USD
Continue using only the good quotes?

Not sure what you mean by:
  If you can summarise various sources -> fq sources let me know.
Sources are detailed at https://code.gnucash.org/docs/C/gnucash-help/fq-sources.html

goodvibes2 commented:
@christopherlam , I understand now why it's trying to get a currency conversion rate. Like stock commodities, there is a 'get_quotes' flag on currency commodities which is set by default if there are any accounts with that currency, although it can be turned off in the Security Editor. After I turned that off for my currencies, the missing alphavantage key error correctly no longer occurs.
I'll clean up my commits and force push.

goodvibes2 commented:
> Here's a better patch. It works when ALPHAVANTAGE is empty, but will not replicate the previous behaviour to retrieve partial quotes. It's too difficult to do so.
> 
Hi @christopherlam,
Please explain more re 'but will not replicate the previous behaviour to retrieve partial quotes. It's too difficult to do so.' so I can add to commit msg?

christopherlam commented:
I don't understand price-quotes.scm nor Finance::Quote completely, but I think it supports multiple quote sources, eg Alphavantage, Yahoo, etc. See the long list in the Securities Editor.

The old code will pass them all to F::Q, and F::Q will return the successful ones. price-quotes.scm will analyse them all, and offer to skip invalid ones.

My patch will simply test for the presence of ALPHAVANTAGE env vs alphavantage-sourced commodities; for example if env was unset or API-key invalid, it wouldn't send all fq-call-data to F::Q and process non-alphavantage results. Ideally it would declare that alphavantage-sourced commodites cannot be quoted and process other sources. But in this patch it would simply bail and request env was set correctly. 

This code is written in very convoluted way and I am not going to rewrite in guile nor C. The problem is there is turf war between C/scheme/perl -- which one should do sanity checks? I offered to rewrite all in guile but someone said 'no'. And very soon I will reduce my hacking.

goodvibes2 commented:
Hi Frank, I have changed the links to a new section in the wiki which I added: https://wiki.gnucash.org/wiki/Online_Quotes#Source_Alphavantage.2C_US

I have cleaned up my commits and force pushed, ready for review please.

goodvibes2 commented:
@mtalexander has already committed a fix for the 'use of uninitialized value' problem (which I prefer to mine), so I'll remove that commit from this PR soon.

goodvibes2 commented:
Ready for review again please.

fellen commented on gnucash/price-quotes.scm:
```diff
@@ -447,6 +450,11 @@
 
+     ((memq 'need-alphavantage-key fq-results)
+      (set! keep-going? #f)
+      (show-error (N_ "ERROR: ALPHAVANTAGE_API_KEY must be set for currency \
++ quotes; see https://wiki.gnucash.org/wiki/Online_Quotes#Source_Alphavantage.2C_US")))
```
1, Can you put the URL as parameter in the translatable string? In the current form changing the URL would break all translations.
2. I know you spent much time to update wiki:Online_Quotes. But I still think a reference to the locally installed docs would be the better choice. Imagine, you were living in the outback and had only a very expensive, unstable internet connection.

jralls replied:
>locally installed docs
That would be nice but the only reference is in [Help Tips](https://code.gnucash.org/docs/C/gnucash-help/fq-sources.html) in a form that isn't readily linkable.

goodvibes2 replied:
Hi @fellen, I think the Online Quotes wiki is better because it has more info and it is all on the one page. The Help and the Guide are out of date. To use the manuals as is, you would have to search (not available in the web interface AFAICT) and jump all over the place. If you lived in the outback and have an unstable internet connection, you wouldn't be able to download the latest pdf/mobi manuals and probably would be unable to use online quotes. I'm not interested in maintaining the same documentation in 2 places...

jralls replied:
@goodvibes2 That's pretty much completely wrong. I already supplied the link to the exact page in the Help, so no searching required and it's up-to-date. It would be better still if the page had an anchor for the Alphavantage entry, hence my comment. The Mac and Windows bundles come with the Help and Guide built-in and it's a one-package addition on those Linux distros that don't include it as part of the main package. Gtk has a built-in context help mechanism that already uses the Help, click any of the Help buttons scattered about the program to see it in action. 

Your not being interested in maintaining the documentation is not a valid argument for not using it.

jralls commented:
Merged to make the freeze-string deadline, parameterized the error string in the merge commit.
Thanks, Chris.
